nvflare/app_common/executors/launcher_executor.py

```diff
@@ -55,181 +52,154 @@ def __init__(
 
         # flags to indicate whether the launcher side will send back trained model and/or metrics
         self._train_with_evaluation = train_with_evaluation
         self._train_task_name = train_task_name
         self._evaluate_task_name = evaluate_task_name
         self._submit_model_task_name = submit_model_task_name
-        self._result_fl_model = None
-        self._result_metrics = None
 
         self._from_nvflare_converter_id = from_nvflare_converter_id
         self._from_nvflare_converter: Optional[ParamsConverter] = None
         self._to_nvflare_converter_id = to_nvflare_converter_id
         self._to_nvflare_converter: Optional[ParamsConverter] = None
 
+        self._monitor_launcher_thread = None
+        self._abort_signal = None
+        self._current_task = None
+
     def initialize(self, fl_ctx: FLContext) -> None:
         self._init_launcher(fl_ctx)
         self._init_converter(fl_ctx)
-        self._init_pipe(fl_ctx)
+        self._monitor_launcher_thread = threading.Thread(target=self._monitor_launcher, args=(fl_ctx,), daemon=True)
+        self._monitor_launcher_thread.start()
+
+    def finalize(self, fl_ctx: FLContext) -> None:
+        self._execute_launcher_method_in_thread_executor(method_name="finalize", fl_ctx=fl_ctx)
 
     def handle_event(self, event_type: str, fl_ctx: FLContext) -> None:
         if event_type == EventType.START_RUN:
+            super().handle_event(event_type, fl_ctx)
             self.initialize(fl_ctx)
         elif event_type == EventType.END_RUN:
             if self.launcher is None:
                 raise RuntimeError("Launcher is None.")
             self._job_end = True
-            self.launcher.finalize(fl_ctx)
-            self.log_info(fl_ctx, "END_RUN event received - telling external to stop")
-            if self.pipe_handler is not None:
-                self.pipe_handler.notify_end("END_RUN event received")
-                self.pipe_handler.stop(close_pipe=True)
+            if self._abort_signal is not None:
+                self._abort_signal.trigger(f"{EventType.END_RUN} event received - telling external to stop")
+            self.finalize(fl_ctx)
+            self.log_info(fl_ctx, f"{EventType.END_RUN} event received - telling external to stop")
+            super().handle_event(event_type, fl_ctx)
 
     def execute(self, task_name: str, shareable: Shareable, fl_ctx: FLContext, abort_signal: Signal) -> Shareable:
         self.log_info(fl_ctx, f"execute for task ({task_name})")
+
+        if not self._initialize_external_execution(task_name, shareable, fl_ctx, abort_signal):
+            return make_reply(ReturnCode.EXECUTION_EXCEPTION)
+
+        if self._from_nvflare_converter is not None:
+            shareable = self._from_nvflare_converter.process(shareable, fl_ctx)
+
+        self.resume_pipe_handler()
+        result = super().execute(task_name, shareable, fl_ctx, abort_signal)
+        self.pause_pipe_handler()
```

Comment:
I don't think you should pause pipe handler unconditionally. You should do so only if the launch mode is per-task.
If the mode is launch once, the FlareAgent in 3rd-party will be running and expecting heartbeat from TaskExchanger. But if HB is stopped since pipe handler is paused here, the FlareAgent won't be happy.

Reply:
Got it, I was just disabling the checking of HB.

Changed to stop/resume both checking and sending HB.
And change call to these "resume_pipe_handler" and "pause_pipe_handler" accordingly.

Note that I move the "launch_once" option into the SubprocessLauncher because I feel that whether the external system should launch once or multiple times it depends on the nature of the Launcher.

The LauncherExecutor now will ALWAYS invoke the Launcher's launch_task.

The Launcher implementation should decide whether it wants to start a new process for each task, or start a new process in the very beginning, and each task come it just ignore.
Or it can say every time a task come, it submits a new "task" to it's 3rd party system.